File: LeetCode/Easy/0414-third-maximum-number/0414-third-maximum-number.java
class Solution {
    public int thirdMax(int[] nums) {

        Set<Integer> set = new HashSet<>();
        
        int i = 0;
        int in = 0;
        int max = Integer.MIN_VALUE;
        int middle = Integer.MIN_VALUE;
        int min = Integer.MIN_VALUE;

        while(i < nums.length) {

            if(set.contains(nums[i])) {
                i++;
                continue;
            }

            set.add(nums[i]);

            if (nums[i] > max) {
                min = middle;
                middle = max;
                max = nums[i];
                in++;
            }
            else if (nums[i] > middle) {
                min = middle;
                middle = nums[i];
                in++;
            }
            else if (nums[i] >= min) {
                min = nums[i];
                in++;
            }

            i++;
        }

        if (in >= 3) {
            return min;
        } else {
            return max;
        }
    }
}

File: LeetCode/Easy/0414-third-maximum-number/0414-third-maximum-number.kt
class Solution {
    fun thirdMax(nums: IntArray): Int {

        val dupCheck = mutableSetOf<Int>()

        var check = 0;
        val ans = intArrayOf(-2_147_483_648, -2_147_483_648, -2_147_483_648)

        for(no in nums) {
            if (dupCheck.contains(no)) {
                continue
            }

            dupCheck.add(no)

            if(ans[0] < no) {
                ans[2] = ans[1]
                ans[1] = ans[0]
                ans[0] = no
                check++
            } else if (ans[1] < no) {
                ans[2] = ans[1]
                ans[1] = no
                check++
            } else if (ans[2] <= no) {
                ans[2] = no
                check++
            }
        }

        if (check >= 3) return ans[2]
        return ans[0]
    }
}